projects/concurrency/main_test.go

```diff
@@ -0,0 +1,184 @@
+package main
+
+import (
+	"fmt"
+	"sync"
+	"testing"
+)
+
+func TestCache(t *testing.T) {
+	t.Run("try to get a key from an empty cache", func(t *testing.T) {
+		cache := NewCache[string, int](3)
+
+		_, ok := cache.Get("a")
+
+		if ok != false {
+			t.Errorf("found a key in the cache where there should have been none")
+		}
+	})
+
+	t.Run("add a single key/value to an empty cache", func(t *testing.T) {
+		cache := NewCache[string, int](3)
+
+		insertKey := "a"
+		insertValue := 100
+
+		ok := cache.Put(insertKey, insertValue)
+
+		if ok != false {
+			t.Errorf("found an existing value on the key in the cache, where there should have been no existing value")
+		}
+
+		cacheListElementPointer, ok := cache.entries[insertKey]
```

Comment:
In general we prefer to test against the public API of our types where possible (i.e. calling `cache.Get` rather than looking in `cache.entries`) unless there's a specific internal property of the type we're looking to assert on.

This has a few advantages:
* If we change the internals of the type (e.g. switching from a linked list to some other data structure, or changing the internals of `CacheValue`) the test will still be valid. It's generally _really_ nice if we can make internal changes to an implementation detail without needing to change existing tests, it helps us have confidence we haven't changed the behaviours the tests were there to test.
* Serves as an example of how the type is intended to be used
* Often easier to verify that the tests are correctly written, because the test is all written at one layer of abstraction rather than crossing an abstraction layer.